odo/backends/sql.py

```diff
@@ -103,7 +103,7 @@
     mssql.UNIQUEIDENTIFIER: string,
     # The SQL Server TIMESTAMP value doesn't correspond to the ISO Standard
     # It is instead just a binary(8) value with no relation to dates or times
-    mssql.TIMESTAMP: bytes_,
+    # mssql.TIMESTAMP: bytes_,
```

Comment:
@thequackdaddy what do you think?

Reply:
@postelrich Sorry I looked at this for a couple hours and forgot to write a synopsis. 

I don't fully grasp how this *should* work and there's lots of python Jedi mind tricks in the code doing things I don't fully understand. I think you might also need a [`load_dialect_impl()`](http://docs.sqlalchemy.org/en/latest/core/custom_types.html#sqlalchemy.types.TypeDecorator.load_dialect_impl) function but on my first (admittedly feeble) attempt I failed to get it to work just right. I think that this should return the `TIMESTAMP` type for all dialects other than MSSQL and `VARCHAR` for MSSQL. 

SQL Alchemy is a little over my head at the moment... I can give it another shot.

Reply:
@thequackdaddy I did some playing around and think I figured it out. It looks like the sqlalchemy dialects have a class attribute called `ischema_names` that is a map between the database type and the sqlalchemy type. You can set it to whatever custom type you want. I made the following example. If you do that for the timestamp type in mssql, then odo should appropriately only pick up bits for that.

```
In [1]: import sqlalchemy

In [2]: from sqlalchemy.dialects.sqlite.base import SQLiteDialect

In [3]: class MyInt(sqlalchemy.INTEGER):
   ...:     pass

In [4]: SQLiteDialect.ischema_names['INTEGER'] = MyInt

In [5]: SQLiteDialect.ischema_names['INT'] = MyInt

In [7]: db = sqlalchemy.create_engine('sqlite:///tutorial.db')

In [8]: meta = sqlalchemy.MetaData()

In [9]: meta.reflect(bind=db)

In [10]: users = meta.tables['users']

In [11]: users.columns.values()
Out[11]:
[Column('user_id', MyInt(), table=<users>, primary_key=True, nullable=False),
 Column('name', VARCHAR(length=40), table=<users>),
 Column('age', MyInt(), table=<users>),
 Column('password', VARCHAR(), table=<users>)]
```

Reply:
@postelrich Ahhh this makes sense to me. Should be cake.

Reply:
@postelrich @thequackdaddy the following diff fixes the error at the end of this comment (which is occurring on the pairing of blaze master, https://github.com/blaze/blaze/commit/45f7e40cda292c1932832475c0092ef37ccdf2a8, and odo master,  https://github.com/blaze/odo/commit/39bf5770935c7cb50ec400aec378857379ba88bdat, the time of this writing), by using the type replacement prescribed above.

The error below is failing because a bytes object is discovered instead of a datetime. as  @thequackdaddy describes in https://github.com/blaze/odo/issues/567 and https://github.com/blaze/blaze/pull/1656

I can begin work on creating a failing test case in `test_postgres` or `test_sql` so that we can catch a regression independent of the blaze test suite.

Diff:
```
index bd45606..5d401b3 100644
--- a/odo/backends/sql.py
+++ b/odo/backends/sql.py
@@ -83,6 +83,11 @@ types = {
 
 revtypes = dict(map(reversed, types.items()))
 
+class MSSQLTimestamp(sa.TIMESTAMP):
+    pass
+
+mssql.base.ischema_names['TIMESTAMP'] = MSSQLTimestamp
+
 revtypes.update({
     sa.DATETIME: datetime_,
     sa.TIMESTAMP: datetime_,
@@ -103,7 +108,7 @@ revtypes.update({
     mssql.UNIQUEIDENTIFIER: string,
     # The SQL Server TIMESTAMP value doesn't correspond to the ISO Standard
     # It is instead just a binary(8) value with no relation to dates or times
-    mssql.TIMESTAMP: bytes_,
+    MSSQLTimestamp: bytes_,
 })
 
 # interval types are special cased in discover_typeengine so remove them from
```

Failure in blaze which this fixes:
```
=============================================== FAILURES ===============================================
______________________________ test_date_grouper_repeats_not_one_point_oh ______________________________

    @pytest.mark.skipif('1.0.0' <= LooseVersion(sa.__version__) <= '1.0.1',
                        reason=("SQLAlchemy generates different code in 1.0.0"
                                " and 1.0.1"))
    def test_date_grouper_repeats_not_one_point_oh():
        columns = [sa.Column('amount', sa.REAL),
                   sa.Column('ds', sa.TIMESTAMP)]
        dta = sa.Table('t', sa.MetaData(), *columns)
        t = symbol('t', discover(dta))
>       expr = by(t.ds.year, avg_amt=t.amount.mean())

blaze/compute/tests/test_sql_compute.py:1879: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = t.ds, key = 'year'

    def __getattr__(self, key):
        assert key != '_hash', \
            '%s should set _hash in _init' % type(self).__name__
        try:
>           result = object.__getattribute__(self, key)
E           AttributeError: 'Field' object has no attribute 'year'

blaze/expr/expressions.py:232: AttributeError
```

Reply:
I edited the previous comment to make `MSSQLTimestamp` a subclass of `mssql.TIMESTAMP` instead of `sa.TIMESTAMP`.

Reply:
https://github.com/blaze/odo/pull/589